core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1431,21 +1434,28 @@ private[spark] class DAGScheduler(
                   // If the whole job has finished, remove it
                   if (job.numFinished == job.numPartitions) {
                     markStageAsFinished(resultStage)
-                    cleanupStateForJobAndIndependentStages(job)
-                    try {
-                      // killAllTaskAttempts will fail if a SchedulerBackend does not implement
-                      // killTask.
-                      logInfo(s"Job ${job.jobId} is finished. Cancelling potential speculative " +
-                        "or zombie tasks for this job")
-                      // ResultStage is only used by this job. It's safe to kill speculative or
-                      // zombie tasks in this stage.
-                      taskScheduler.killAllTaskAttempts(
-                        stageId,
-                        shouldInterruptTaskThread(job),
-                        reason = "Stage finished")
-                    } catch {
-                      case e: UnsupportedOperationException =>
-                        logWarning(s"Could not cancel tasks for stage $stageId", e)
+                    val removedStages = cleanupStateForJobAndIndependentStages(job)
```

Comment:
there are other places cleanupStateForJobAndIndependentStages is called, this isn't handling this same thing. I'm wondering if the markStageAsFinished and canceltasks shouldn't be done inside the cleanupStatgeForJobAndIndependentStages

Reply:
ok looking a bit more at the places calling into cleanupStateForJobAndIndependentStages I see that failJobAndIndependentStages has code to canceltasks and markasfinished before calling into cleanupStateForJobAndIndependentStages, although that logic seems a bit broken as well because it only calls cleanupStateForJobAndIndependentStages if it was able to cancel the tasks. which is ok since its not removing it from runningStages so when it actually finishes it should send the message.